ntpd/src/daemon/keyexchange.rs

```diff
@@ -647,4 +660,243 @@ mod tests {
+
+    async fn send_records_to_server(records: Vec<NtsRecord>) -> Result<(), KeyExchangeError> {
+        let listener = TcpListener::bind(&("localhost", 0)).await?;
+        let port = listener.local_addr()?.port();
+
+        tokio::spawn(async move {
+            let mut stream = client_tls_stream("localhost", port).await;
+
+            for record in records {
+                let mut buffer = Vec::with_capacity(1024);
+                record.write(&mut buffer).unwrap();
+
+                stream.write_all(&buffer).await.unwrap();
+            }
+
+            let mut buf = [0; 1024];
+            loop {
+                match stream.read(&mut buf).await.unwrap() {
+                    0 => break,
+                    _ => continue,
+                }
+            }
+        });
+
+        run_server(listener).await
+    }
+
+    #[tokio::test]
+    async fn receive_cookies() {
+        let result = send_records_to_client(vec![
+            NtsRecord::NextProtocol {
+                protocol_ids: vec![0],
+            },
+            NtsRecord::AeadAlgorithm {
+                critical: false,
+                algorithm_ids: vec![15],
+            },
+            NtsRecord::NewCookie {
+                cookie_data: vec![1, 2, 3],
+            },
+            NtsRecord::EndOfMessage,
+        ])
+        .await;
+
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
+    async fn records_after_end_are_ignored() {
+        let result = send_records_to_client(vec![
+            NtsRecord::NextProtocol {
+                protocol_ids: vec![0],
+            },
+            NtsRecord::AeadAlgorithm {
+                critical: false,
+                algorithm_ids: vec![15],
+            },
+            NtsRecord::NewCookie {
+                cookie_data: vec![1, 2, 3],
+            },
+            NtsRecord::EndOfMessage,
+            NtsRecord::NewCookie {
+                cookie_data: vec![1, 2, 3],
+            },
+        ])
+        .await;
+
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
+    async fn no_cookies() {
+        let result = send_records_to_client(vec![NtsRecord::EndOfMessage]).await;
+
+        let error = result.unwrap_err();
+
+        assert!(matches!(error, KeyExchangeError::NoCookies));
+    }
+
+    #[tokio::test]
+    async fn server_expected_client_records() {
+        let records = NtsRecord::client_key_exchange_records().to_vec();
+        let result = send_records_to_server(records).await;
+
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
+    async fn immediate_end_of_message() {
+        let records = vec![NtsRecord::EndOfMessage];
+        let result = send_records_to_server(records).await;
+
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
+    async fn double_end_of_message() {
+        let records = vec![NtsRecord::EndOfMessage, NtsRecord::EndOfMessage];
+        let result = send_records_to_server(records).await;
+
+        assert!(result.is_ok());
+    }
```

Comment:
it's unclear if this is really correct according to the spec. Fine in practice assuming a valid client.